ssh/tailssh/tailssh.go

```diff
@@ -122,19 +125,37 @@ func (srv *server) newSSHServer() (*ssh.Server, error) {
 		},
 		Version:                     "SSH-2.0-Tailscale",
 		LocalPortForwardingCallback: srv.mayForwardLocalPortTo,
-		NoClientAuthCallback: func(m gossh.ConnMetadata) (*gossh.Permissions, error) {
-			if srv.requiresPubKey(m.User(), toIPPort(m.LocalAddr()), toIPPort(m.RemoteAddr())) {
-				return nil, errors.New("public key required") // any non-nil error will do
+		NoClientAuthCallback: func(conn gossh.ConnMetadata) (*gossh.Permissions, error) {
+			var err error
+			ci, err = srv.connInfo(conn.User(), toIPPort(conn.LocalAddr()), toIPPort(conn.RemoteAddr()))
+			if err != nil {
+				srv.logf("failed to get conninfo: %v", err)
+				return nil, gossh.ErrDenied
+			}
+			if _, withoutPubKey := srv.canProceed(ci, nil); withoutPubKey {
```

Comment:
not a huge fan of all these new ways of spelling "ok" you're adding like "yes" and "withoutPubKey". This one in particular is very verbose.

You could write "ok" instead and then put "// Can proceed without a public key" or `canProceed(ci, nil /* without pubkey */)`